
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/introduction-to-shell/Supplementary/1-supplementary/">
      
      
        <link rel="prev" href="../../06-organization/">
      
      
      
      <link rel="icon" href="../../fig/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>1. Supplementary - File manipulation - Introduction to  Shell for Bioinformatics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../../assets/javascripts/glightbox.min.js"></script></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#1-supplementary-file-manipulation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Introduction to  Shell for Bioinformatics" class="md-header__button md-logo" aria-label="Introduction to  Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../../fig/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to  Shell for Bioinformatics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1. Supplementary - File manipulation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/introduction-to-shell" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/introduction-to-shel
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Introduction to  Shell for Bioinformatics" class="md-nav__button md-logo" aria-label="Introduction to  Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../../fig/nesi_ga.png" alt="logo">

    </a>
    Introduction to  Shell for Bioinformatics
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/introduction-to-shell" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/introduction-to-shel
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../0-unixl-linux-overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview : UNIX, Linux &amp; UNIX Shell
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../01-introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1. First steps - Say Goodbye to the 🐭 ( mostly)
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../02-the-filesystem/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2. Navigating Files and Directories
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../03-working-with-files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3. Working with Files and Directories
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../04-redirection/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4. Redirection
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../05-writing-scripts/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5. Writing Scripts and Working with Data
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../06-organization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6. Project Organization
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" checked>
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Supplementary
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    1. Supplementary - File manipulation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    1. Supplementary - File manipulation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#file-manipulation-and-more-practice-with-pipes" class="md-nav__link">
    <span class="md-ellipsis">
      File manipulation and more practice with pipes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="File manipulation and more practice with pipes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-many-of-the-read-libraries-are-paired-end" class="md-nav__link">
    <span class="md-ellipsis">
      How many of the read libraries are paired end?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#solution" class="md-nav__link">
    <span class="md-ellipsis">
      Solution
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Solution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-many-of-each-class-of-library-layout-are-there" class="md-nav__link">
    <span class="md-ellipsis">
      How many of each class of library layout are there?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file" class="md-nav__link">
    <span class="md-ellipsis">
      Can we sort the file by library layout and save that sorted information to a new file?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#can-we-extract-only-paired-end-records-into-a-new-file" class="md-nav__link">
    <span class="md-ellipsis">
      Can we extract only paired-end records into a new file?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise_1" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#solution_1" class="md-nav__link">
    <span class="md-ellipsis">
      Solution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-code-more-customizeable-using-command-line-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      Making code more customizeable using command line arguments
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#file-manipulation-and-more-practice-with-pipes" class="md-nav__link">
    <span class="md-ellipsis">
      File manipulation and more practice with pipes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="File manipulation and more practice with pipes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-many-of-the-read-libraries-are-paired-end" class="md-nav__link">
    <span class="md-ellipsis">
      How many of the read libraries are paired end?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#solution" class="md-nav__link">
    <span class="md-ellipsis">
      Solution
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Solution">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-many-of-each-class-of-library-layout-are-there" class="md-nav__link">
    <span class="md-ellipsis">
      How many of each class of library layout are there?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file" class="md-nav__link">
    <span class="md-ellipsis">
      Can we sort the file by library layout and save that sorted information to a new file?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#can-we-extract-only-paired-end-records-into-a-new-file" class="md-nav__link">
    <span class="md-ellipsis">
      Can we extract only paired-end records into a new file?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise_1" class="md-nav__link">
    <span class="md-ellipsis">
      Exercise
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#solution_1" class="md-nav__link">
    <span class="md-ellipsis">
      Solution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-code-more-customizeable-using-command-line-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      Making code more customizeable using command line arguments
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="1-supplementary-file-manipulation">1. Supplementary - File manipulation<a class="headerlink" href="#1-supplementary-file-manipulation" title="Permanent link">&para;</a></h1>
<h2 id="file-manipulation-and-more-practice-with-pipes">File manipulation and more practice with pipes<a class="headerlink" href="#file-manipulation-and-more-practice-with-pipes" title="Permanent link">&para;</a></h2>
<p>Let's use the tools we've added to our tool kit so far, along with a few new ones, to example our SRA metadata file. First, let's navigate to the correct directory.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>$<span class="w"> </span><span class="nb">cd</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>$<span class="w"> </span><span class="nb">cd</span><span class="w"> </span>~/shell_data/sra_metadata
</code></pre></div>
<p>This file contains a lot of information about the samples that we submitted for sequencing. We
took a look at this file in an earlier lesson. Here we're going to use the information in this
file to answer some questions about our samples.</p>
<h4 id="how-many-of-the-read-libraries-are-paired-end">How many of the read libraries are paired end?<a class="headerlink" href="#how-many-of-the-read-libraries-are-paired-end" title="Permanent link">&para;</a></h4>
<p>The samples that we submitted to the sequencing facility were a mix of single and paired end
libraries. We know that we recorded information in our metadata table about which samples used
which library preparation method, but we don't remember exactly where this data is recorded.
Let's start by looking at our column headers to see which column might have this information. Our
column headers are in the first row of our data table, so we can use <code>head</code> with a <code>-n</code> flag to
look at just the first row of the file.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>$<span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>SraRunTable.txt
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="go">BioSample_s InsertSize_l    LibraryLayout_s Library_Name_s  LoadDate_s  MBases_l    MBytes_l    ReleaseDate_s Run_s SRA_Sample_s Sample_Name_s Assay_Type_s AssemblyName_s BioProject_s Center_Name_s Consent_s Organism_Platform_s SRA_Study_s g1k_analysis_group_s g1k_pop_code_s source_s strain_s</span>
</code></pre></div>
<p>That is only the first line of our file, but because there are a lot of columns, the output
likely wraps around your terminal window and appears as multiple lines. Once we figure out which
column our data is in, we can use a command called <code>cut</code> to extract the column of interest.</p>
<p>Because this is pretty hard to read, we can look at just a few column header names at a time by combining the <code>|</code> redirect and <code>cut</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>$<span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f1-4
</code></pre></div>
<p><code>cut</code> takes a <code>-f</code> flag, which stands for "field". This flag accepts a list of field numbers,
in our case, column numbers. Here we are extracting the first four column names.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="go">BioSample_s InsertSize_l      LibraryLayout_s   Library_Name_s    </span>
</code></pre></div>
<p>The LibraryLayout_s column looks like it should have the information we want.  Let's look at some
of the data from that column. We can use <code>cut</code> to extract only the 3<sup>rd</sup> column from the file and
then use the <code>|</code> operator with <code>head</code> to look at just the first few lines of data in that column.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="go">LibraryLayout_s</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a><span class="go">SINGLE</span>
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a><span class="go">PAIRED</span>
</code></pre></div>
<p>We can see that there are (at least) two categories, SINGLE and PAIRED.  We want to search all entries in this column
for just PAIRED and count the number of matches. For this, we will use the <code>|</code> operator twice
to combine <code>cut</code> (to extract the column we want), <code>grep</code> (to find matches) and <code>wc</code> (to count matches).</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>PAIRED<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="go">2</span>
</code></pre></div>
<p>We can see from this that we have only two paired-end libraries in the samples we submitted for
sequencing.</p>
<p>:::::::::::::::::::::::::::::::::::::::  challenge</p>
<h2 id="exercise">Exercise<a class="headerlink" href="#exercise" title="Permanent link">&para;</a></h2>
<p>How many single-end libraries are in our samples?</p>
<p>:::::::::::::::  solution</p>
<h2 id="solution">Solution<a class="headerlink" href="#solution" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>SINGLE<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="go">35</span>
</code></pre></div>
<h4 id="how-many-of-each-class-of-library-layout-are-there">How many of each class of library layout are there?<a class="headerlink" href="#how-many-of-each-class-of-library-layout-are-there" title="Permanent link">&para;</a></h4>
<p>We can extract even more information from our metadata table if we add in some new tools: <code>sort</code> and <code>uniq</code>. The <code>sort</code> command will sort the lines of a text file and the <code>uniq</code> command will
filter out repeated neighboring lines in a file. You might expect <code>uniq</code> to
extract all of the unique lines in a file. This isn't what it does, however, for reasons
involving computer memory and speed. If we want to extract all unique lines, we
can do so by combining <code>uniq</code> with <code>sort</code>. We'll see how to do this soon.</p>
<p>For example, if we want to know how many samples of each library type are recorded in our table,
we can extract the third column (with <code>cut</code>), and pipe that output into <code>sort</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort
</code></pre></div>
<p>If you look closely, you might see that we have one line that reads "LibraryLayout_s". This is the
header of our column. We can discard this information using the <code>-v</code> flag in <code>grep</code>, which means
return all the lines that <strong>do not</strong> match the search pattern.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>LibraryLayout_s<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort
</code></pre></div>
<p>This command returns a sorted list (too long to show here) of PAIRED and SINGLE values. We can use
the <code>uniq</code> command to see a list of all the different categories that are present. If we do this,
we see that the only two types of libraries we have present are labelled PAIRED and SINGLE. There
aren't any other types in our file.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>LibraryLayout_s<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="go">PAIRED</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="go">SINGLE</span>
</code></pre></div>
<p>If we want to count how many of each we have, we can use the <code>-c</code> (count) flag for <code>uniq</code>.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>$<span class="w"> </span>cut<span class="w"> </span>-f3<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>LibraryLayout_s<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="go">2 PAIRED</span>
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="go">35 SINGLE</span>
</code></pre></div>
<div class="admonition dumbbell">
<p class="admonition-title">Exercise</p>
<ol>
<li>How many different sample load dates are there?</li>
<li>How many samples were loaded on each date?</li>
</ol>
<details class="success">
<summary>Solution</summary>
<ol>
<li>There are two different sample load dates.</li>
</ol>
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>cut<span class="w"> </span>-f5<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>LoadDate_s<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="go">25-Jul-12</span>
<a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="go">29-May-14</span>
</code></pre></div>
<ol>
<li>Six samples were loaded on one date and 31 were loaded on the other.</li>
</ol>
<div class="highlight"><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>cut<span class="w"> </span>-f5<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span>LoadDate_s<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a><span class="go"> 6 25-Jul-12</span>
<a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a><span class="go">31 29-May-14</span>
</code></pre></div>
</details>
</div>
<h4 id="can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file">Can we sort the file by library layout and save that sorted information to a new file?<a class="headerlink" href="#can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file" title="Permanent link">&para;</a></h4>
<p>We might want to re-order our entire metadata table so that all of the paired-end samples appear
together and all of the single-end samples appear together. We can use the <code>-k</code> (key) flag for <code>sort</code> to
sort based on a particular column. This is similar to the <code>-f</code> flag for <code>cut</code>.</p>
<p>Let's sort based on the third column (<code>-k3</code>) and redirect our output to a new file.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>$<span class="w"> </span>sort<span class="w"> </span>-k3<span class="w"> </span>SraRunTable.txt<span class="w"> </span>&gt;<span class="w"> </span>SraRunTable_sorted_by_layout.txt
</code></pre></div>
<h4 id="can-we-extract-only-paired-end-records-into-a-new-file">Can we extract only paired-end records into a new file?<a class="headerlink" href="#can-we-extract-only-paired-end-records-into-a-new-file" title="Permanent link">&para;</a></h4>
<p>We also might want to extract the information for all samples that meet a specific criterion
(for example, are paired-end) and put those lines of our table in a new file. First, we need
to check to make sure that the pattern we're searching for ("PAIRED") only appears in the column
where we expect it to occur (column 3). We know from earlier that there are only two paired-end
samples in the file, so we can <code>grep</code> for "PAIRED" and see how many results we get.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>$<span class="w"> </span>grep<span class="w"> </span>PAIRED<span class="w"> </span>SraRunTable.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a><span class="go">2</span>
</code></pre></div>
<p>There are only two results, so we can use "PAIRED" as our search term to extract the paired-end
samples to a new file.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>$<span class="w"> </span>grep<span class="w"> </span>PAIRED<span class="w"> </span>SraRunTable.txt<span class="w"> </span>&gt;<span class="w"> </span>SraRunTable_only_paired_end.txt
</code></pre></div>
<h2 id="exercise_1">Exercise<a class="headerlink" href="#exercise_1" title="Permanent link">&para;</a></h2>
<p>Sort samples by load date and export each of those sets to a new file (one new file per
unique load date).</p>
<h2 id="solution_1">Solution<a class="headerlink" href="#solution_1" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>$<span class="w"> </span>grep<span class="w"> </span><span class="m">25</span>-Jul-12<span class="w"> </span>SraRunTable.txt<span class="w"> </span>&gt;<span class="w"> </span>SraRunTable_25-Jul-12.txt
<a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a>$<span class="w"> </span>grep<span class="w"> </span><span class="m">29</span>-May-14<span class="w"> </span>SraRunTable.txt<span class="w"> </span>&gt;<span class="w"> </span>SraRunTable_29-May-14.txt
</code></pre></div>
<h2 id="making-code-more-customizeable-using-command-line-arguments">Making code more customizeable using command line arguments<a class="headerlink" href="#making-code-more-customizeable-using-command-line-arguments" title="Permanent link">&para;</a></h2>
<p>In Lesson 05 (Writing Scripts) we used the <code>grep</code> command line tool to look for FASTQ records with
lots of Ns from all the .fastq files in our current folder using the following code:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>$<span class="w"> </span>grep<span class="w"> </span>-B1<span class="w"> </span>-A2<span class="w"> </span>-h<span class="w"> </span>NNNNNNNNNN<span class="w"> </span>*.fastq<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s1">&#39;^--&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>scripted_bad_reads.txt
</code></pre></div>
<p>This is very useful, but could be more customizeable. We may want to be able to run this command
over and over again without needing to copy and paste it and allow the user to specify exactly which
file they want to examine for bad reads.</p>
<p>We can accomplish these goals by including the above command in a script that takes in user input
via a command line argument. We can slightly modify our <code>bad-reads-script.sh</code> file to do so. Use
<code>c</code> to copy your <code>bad-reads-script.sh</code> into a new script called <code>custom-bad-reads-script.sh</code>. Make
the following modifications to <code>custom-bad-reads-script.sh</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a><span class="nv">filename</span><span class="o">=</span><span class="nv">$1</span>
<a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a>grep<span class="w"> </span>-B1<span class="w"> </span>-A2<span class="w"> </span>-h<span class="w"> </span>NNNNNNNNNN<span class="w"> </span><span class="nv">$filename</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s1">&#39;^--&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>scripted_bad_reads.txt
</code></pre></div>
<p><code>$1</code> is our command line argument. The line <code>filename=$1</code> tells Bash to take the first thing you type
after the name of the script itself and assign that value to a variable called filename.</p>
<p>For example, this script can be run in the following way to output the bad reads just from one file:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>bash<span class="w"> </span>custom-bad-reads-script.sh<span class="w"> </span>SRR098026.fastq
</code></pre></div>
<p>We can then take a look at what the output file currently contains using <code>head scripted_bad_reads.txt</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a><span class="go">@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35</span>
<a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a><span class="go">NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN</span>
<a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a><span class="go">+SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35</span>
<a id="__codelineno-29-4" name="__codelineno-29-4" href="#__codelineno-29-4"></a><span class="go">!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!</span>
<a id="__codelineno-29-5" name="__codelineno-29-5" href="#__codelineno-29-5"></a><span class="go">@SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35</span>
<a id="__codelineno-29-6" name="__codelineno-29-6" href="#__codelineno-29-6"></a><span class="go">NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN</span>
<a id="__codelineno-29-7" name="__codelineno-29-7" href="#__codelineno-29-7"></a><span class="go">+SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35</span>
<a id="__codelineno-29-8" name="__codelineno-29-8" href="#__codelineno-29-8"></a><span class="go">!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!</span>
<a id="__codelineno-29-9" name="__codelineno-29-9" href="#__codelineno-29-9"></a><span class="go">@SRR098026.3 HWUSI-EAS1599_1:2:1:0:570 length=35</span>
<a id="__codelineno-29-10" name="__codelineno-29-10" href="#__codelineno-29-10"></a><span class="go">NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN</span>
</code></pre></div>
<p>This should be same output as using our original code and manually modifying the
original standalone code on the command line to "SRR098026.fastq" on the command line,
which should give us the same output as above:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>$<span class="w"> </span>grep<span class="w"> </span>-B1<span class="w"> </span>-A2<span class="w"> </span>-h<span class="w"> </span>NNNNNNNNNN<span class="w"> </span>SRR098026.fastq<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s1">&#39;^--&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>scripted_bad_reads.txt
<a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a>head<span class="w"> </span>scripted_bad_reads.txt
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a><span class="go">@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35</span>
<a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a><span class="go">NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN</span>
<a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a><span class="go">+SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35</span>
<a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a><span class="go">!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!</span>
<a id="__codelineno-31-5" name="__codelineno-31-5" href="#__codelineno-31-5"></a><span class="go">@SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35</span>
<a id="__codelineno-31-6" name="__codelineno-31-6" href="#__codelineno-31-6"></a><span class="go">NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN</span>
<a id="__codelineno-31-7" name="__codelineno-31-7" href="#__codelineno-31-7"></a><span class="go">+SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35</span>
<a id="__codelineno-31-8" name="__codelineno-31-8" href="#__codelineno-31-8"></a><span class="go">!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!</span>
<a id="__codelineno-31-9" name="__codelineno-31-9" href="#__codelineno-31-9"></a><span class="go">@SRR098026.3 HWUSI-EAS1599_1:2:1:0:570 length=35</span>
<a id="__codelineno-31-10" name="__codelineno-31-10" href="#__codelineno-31-10"></a><span class="go">NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN</span>
</code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../../06-organization/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 6. Project Organization">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                6. Project Organization
              </div>
            </div>
          </a>
        
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Introduction Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  <script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(() => { lightbox.reload() });
</script></body>
</html>